lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +260,479 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+    SmallVector<SILValue, 8> args =
+        createArgsArray(branch->getArgs(), /*appendNewArgs=*/true);
+    newTermInst =
+        builder.createBranch(branch->getLoc(), branch->getDestBB(), args);
+  } else if (auto *condBranch = dyn_cast<CondBranchInst>(termInst)) {
+    bool isTrueEdge = condBranch->getTrueBB() == target;
+    assert(((isTrueEdge && condBranch->getTrueBB() == target) ||
+            (!isTrueEdge && condBranch->getFalseBB() == target)) &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> trueArgs = createArgsArray(
+        condBranch->getTrueArgs(), /*appendNewArgs=*/isTrueEdge);
+    SmallVector<SILValue, 8> falseArgs = createArgsArray(
+        condBranch->getFalseArgs(), /*appendNewArgs=*/!isTrueEdge);
+    newTermInst = builder.createCondBranch(
+        condBranch->getLoc(), condBranch->getCondition(),
+        condBranch->getTrueBB(), trueArgs, condBranch->getFalseBB(), falseArgs,
+        condBranch->getTrueBBCount(), condBranch->getFalseBBCount());
+  } else {
+    // At the moment we can only add arguments to br and cond_br.
+    llvm_unreachable("Can't add argument to terminator");
+  }
+  // Remove the old terminator instruction.
+  termInst->dropAllReferences();
+  termInst->eraseFromParent();
+  return newTermInst;
+}
+
+std::pair<SILBasicBlock *, SILValue>
+SingleExitLoopTransformer::createNewHeader() {
+
+  SILBuilder builder(header);
+  ASTContext &context = builder.getASTContext();
+
+  SILBasicBlock *newHeader = currentFn->createBasicBlock();
+  loop->addBasicBlockToLoop(newHeader, LI->getBase()); // Should be done first.
+  // Clone arguments and change all uses to the new header's arguments.
+  newHeader->cloneArgumentList(header);
+  SmallVector<SILValue, 8> headerArgs;
+  for (auto i : indices(header->getArguments())) {
+    header->getArgument(i)->replaceAllUsesWith(newHeader->getArgument(i));
+    headerArgs.push_back(newHeader->getArgument(i));
+  }
+  header->dropAllArguments();
+  // Add phi arguments in the new header corresponding to the escaping values.
+  for (const auto &escapingValue : escapingValues) {
+    SILValue newValue = newHeader->createPHIArgument(
+        escapingValue->getType(), escapingValue.getOwnershipKind());
+    escapingValue->replaceAllUsesWith(newValue);
+  }
+  // An integer to identify the exit edge.
+  SILValue exitIndexArg = newHeader->createPHIArgument(
+      convertElementTypeToTensorValueType(
+          SILType::getBuiltinIntegerType(32, context)),
+      ValueOwnershipKind::Owned);
+  // A boolean corresponding to the stayInLoop flag.
+  newHeader->createPHIArgument(convertElementTypeToTensorValueType(
+                                   SILType::getBuiltinIntegerType(1, context)),
+                               ValueOwnershipKind::Trivial);
+  return std::make_pair(newHeader, exitIndexArg);
+}
+
+SILBasicBlock *
+SingleExitLoopTransformer::createLatchBlock(SILBasicBlock *newHeader) {
```

Comment:
nit: the previous func is called `createNewHeader`. should this be `createNewLatch`?